### PR TITLE
Fix erroneous UserWarning for unset max_epochs

### LINKED ISSUE #13257: Erroneous UserWarning for unset max_epochs
## 🐛 Bug

When I use:
```python
Trainer(fast_dev_run=True)
```
I see the following warning:
```
pytorch_lightning.utilities.warnings.PossibleUserWarning: `max_epochs` was not set. Setting it to 1000 epochs. To train without an epoch limit, set `max_epochs=-1`.
```
However, `fast_dev_run=True` implies `max_epochs=1`, so shouldn't this warning be skipped?

### To Reproduce

Use code like:
```python
Trainer(fast_dev_run=True)
```
in your code with warnings enabled or captured by pytest.

### Expected behavior

No warning message.

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): 1.6.4
- PyTorch Version (e.g., 1.10): 1.11.0
- Python version (e.g., 3.9): 3.10.4, etc.
- OS (e.g., Linux): Linux, etc.
- CUDA/cuDNN version: N/A
- GPU models and configuration: N/A
- How you installed PyTorch (`conda`, `pip`, source): pip
- If compiling from source, the output of `torch.__config__.show()`: N/A
- Any other relevant information: N/A

cc @justusschock @kaushikb11 @awaelchli @borda @rohitgr7

### PR DESCRIPTION
**Pull request by adamjstewart**
## What does this PR do?

Fixes #13257. @akihironitta

Now the warning is raised only during `trainer.fit` call.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/flags/test_min_max_epochs.py [hunk 1]
```diff
@@ -41,3 +43,11 @@ def test_max_epochs_not_set_warning():
     with pytest.warns(PossibleUserWarning, match="`max_epochs` was not set. Setting it to 1000 epochs."):
         trainer = Trainer(max_epochs=None)
         assert trainer.max_epochs == 1000
+
+
+def test_fast_dev_run_no_warning():
+    """Test that no warning is emitted when `fast_dev_run` is set."""
+    with warnings.catch_warnings():
+        warnings.simplefilter("error")
```
Comment (adamjstewart) [inline tests/trainer/flags/test_min_max_epochs.py:51]: See https://docs.pytest.org/en/7.0.x/how-to/capture-warnings.html#additional-use-cases-of-warnings-in-tests

**Comment by adamjstewart** [comment]
Tests and fix were separated into 2 commits to show that the tests do indeed fail without the fix.

File: tests/trainer/flags/test_min_max_epochs.py (hunk 1 above)
Comment (carmocca) [reply tests/trainer/flags/test_min_max_epochs.py:51]: You can use this utility we have:

```
from tests.helpers.utils import no_warning_call

with no_warning_call(match="..."):
    ...
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -449,7 +449,7 @@ def __init__(
         self.tuner = Tuner(self)
 
         min_steps, max_steps, min_epochs, max_epochs, max_time = _parse_loop_limits(
-            min_steps, max_steps, min_epochs, max_epochs, max_time
+            min_steps, max_steps, min_epochs, max_epochs, max_time, fast_dev_run
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:452]: IMO instead of passing `fast_dev_run` here, a better solution would be to call this

https://github.com/PyTorchLightning/pytorch-lightning/blob/bfa8b7be2d99b980afa62f5cb0433326bcfd2ef0/pytorch_lightning/trainer/trainer.py#L544

before `parse_loop_limits` since `fast_dev_run` might modify the limits.

Comment (adamjstewart) [reply pytorch_lightning/trainer/trainer.py]: It looks like moving `_init_debugging_flags` above `_parse_loop_limits` breaks because it requires `fit_loop` to be set, but that is set using values from the output of `_parse_loop_limits`. Should I revert to my previous solution or is there a way to break this loop?

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py]: I pushed a commit that should fix the problem. It required inlining the `init_debugging_flags` method.

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -448,6 +448,59 @@ def __init__(
         self._signal_connector = SignalConnector(self)
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:448]: Can you update the CHANGELOG? "Fixed" section

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — DISMISSED** [review]

**Review by akihironitta — COMMENTED** [review]
@adamjstewart Thank you for working on this! Would you mind having a look at the failing tests in CI? https://github.com/PyTorchLightning/pytorch-lightning/runs/6895684273?check_suite_focus=true

I think applying the following patch should fix the error :)

**Comment by adamjstewart** [comment]
@carmocca it seems like your solution isn't passing the tests

File: tests/tests_pytorch/trainer/flags/test_min_max_epochs.py [hunk 4]
```diff
@@ -37,7 +38,13 @@ def test_min_max_steps_epochs(tmpdir, min_epochs, max_epochs, min_steps, max_ste
...
+    """Test that a warning is only emitted when `max_epochs` was not set by the user."""
+    match = "`max_epochs` was not set. Setting it to 1000 epochs."
+
+    with pytest.warns(PossibleUserWarning, match=match):
         trainer = Trainer(max_epochs=None)
         assert trainer.max_epochs == 1000
+
+    with no_warning_call(expected_warning=PossibleUserWarning, match=match):
+        Trainer(fast_dev_run=True)
+        Trainer(fast_dev_run=1)
```
[the remark this answers is not in the document]
Comment (adamjstewart) [reply tests/tests_pytorch/trainer/flags/test_min_max_epochs.py:50]: We could try 2. I think it's important to try all input types and make sure the parameter behaves as expected. For example, if this PR checked to see `if fast_dev_run == True:` (not a good solution, I know), then this test would catch that mistake.

**Comment by carmocca** [comment]
> it seems like your solution isn't passing the tests

Should be passing now

File: src/pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -536,75 +578,9 @@ def __init__(
...
-            limit_predict_batches,
-            val_check_interval,
-            overfit_batches,
-            fast_dev_run,
-        )
-
         # Callback system
         self._call_callback_hooks("on_init_end")
 
-    def _init_debugging_flags(
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/trainer.py:558]: qq: do you think we could dep & remove `overfit_batches`?

all it does is this:
```py
if overfit_batches_enabled:
    self.limit_train_batches = overfit_batches
    self.limit_val_batches = overfit_batches
```

**Review by akihironitta — APPROVED** [review]
It looks good to me other than the possible readability issue https://github.com/Lightning-AI/lightning/pull/13262/files#r903701888

**Comment by adamjstewart** [comment]
Is this PR ready to merge or is there anything else it needs?

File: src/pytorch_lightning/trainer/trainer.py [hunk 6]
```diff
@@ -456,6 +456,65 @@ def __init__(
...
+            self.limit_predict_batches = _determine_batch_limits(limit_predict_batches, "limit_predict_batches")
+            self.num_sanity_val_steps = float("inf") if num_sanity_val_steps == -1 else num_sanity_val_steps
+            self.val_check_interval = _determine_batch_limits(val_check_interval, "val_check_interval")
+        if overfit_batches_enabled:
+            self.limit_train_batches = overfit_batches
+            self.limit_val_batches = overfit_batches
+
         min_steps, max_steps, min_epochs, max_epochs, max_time = _parse_loop_limits(
             min_steps, max_steps, min_epochs, max_epochs, max_time
         )
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:520]: While `_parse_loop_limits` doesn't do much with it, max_time by construction of the function call could be modified. Previously it was passed to `self._callback_connector.on_trainer_init` after these lines but now it has moved to before.

In other words, before we did the parsing first before using the values, now it is the other way around.

An alternative solution that avoids the re-ordering would have been to pass the `fast_dev_run` flag down to the `_parse_loop_limits` to be able to avoid emitting the warning. Was this considered? If yes, what was the blocker for not being able to do it this way?

Comment (adamjstewart) [reply src/pytorch_lightning/trainer/trainer.py:520]: > An alternative solution that avoids the re-ordering would have been to pass the `fast_dev_run` flag down to the `_parse_loop_limits` to be able to avoid emitting the warning. Was this considered?

This was actually my initial solution. Aside from testing, the entire PR was a single commit: https://github.com/Lightning-AI/lightning/pull/13262/commits/97e73aab50f6192bae29fc5d419c989ded21a85b.

> If yes, what was the blocker for not being able to do it this way?

https://github.com/Lightning-AI/lightning/pull/13262#discussion_r897041894

**Review by rohitgr7 — CHANGES_REQUESTED** [review]
I'd consider implementing https://github.com/Lightning-AI/lightning/pull/13262#issuecomment-1164448685

this way we can keep those methods that handles debugging flags.

**Comment by carmocca** [comment]
> start raising this warning during .fit call cause during that call it actually makes sense

This PR addresses a bug in our existing logic, whereas this proposal would be a refactor and should be left for a follow-up

> do you think we could dep & remove overfit_batches?

Same comment about this suggestion. This should be left for a separate issue where we discuss the advantages and disadvantages.

I don't think it's relevant to keep the separate method because it's not called from anywhere else and it simplifies the attribute setting of the Trainer instance. The code had to be moved around to make sure things are set where expected, but not looking at the diff I feel like the code is somewhat easier to follow now for a new user.

**Comment by adamjstewart** [comment]
I assume these test failures are unrelated to the PR?

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/loops/fit_loop.py [hunk 7]
```diff
@@ -50,10 +50,10 @@ class FitLoop(Loop[None]):
     def __init__(
         self,
         min_epochs: int = 0,
-        max_epochs: int = 1000,
+        max_epochs: Optional[int] = None,
```
Comment (carmocca) [inline src/pytorch_lightning/loops/fit_loop.py:53]: @rohitgr7 Why this change? It's cleaner if the loop takes the argument already validated (converted to int). We should always try to avoid Optionals

Comment (rohitgr7) [reply src/pytorch_lightning/loops/fit_loop.py:53]: the issue is max_epochs doesn't do anything for `trainer.test/validate/predict` but we initialize the fit_loop in the trainer constructor. We need to ensure it gets updated and that warning should be raised during `.fit` only. Due to this, I don't see a possibility of setting it to an int before the call.

**Comment by adamjstewart** [comment]
Is there still interest in this PR? My original commits solved the issue in only a few lines of code and passed all tests. Since then, this PR has blown up into a huge refactor with 4+ contributors, 23 commits, and 9 files changed, and is still failing tests. I no longer have any idea how this code works or if this is the right thing to do. Would it be possible to merge the first couple of commits and leave the refactor for another day?

**Comment by awaelchli** [comment]
@adamjstewart We haven't forgotten about this. I would very much prefer to go with the simple solution (see these two [comments](https://github.com/Lightning-AI/lightning/pull/13262#discussion_r909087735)). Any larger issues about how we should validate some input arguments that only apply when certain trainer functions get called should be left for a different issue in my opinion, as this is a broader question than just fast dev run.

**Review by carmocca — APPROVED** [review]

**Comment by carmocca** [comment]
LGTM. Just needs passing CI and your approval @awaelchli

**Comment by rohitgr7** [comment]
> @adamjstewart We haven't forgotten about this. I would very much prefer to go with the simple solution (see these two [comments](https://github.com/Lightning-AI/lightning/pull/13262#discussion_r909087735)). Any larger issues about how we should validate some input arguments that only apply when certain trainer functions get called should be left for a different issue in my opinion, as this is a broader question than just fast dev run.
> 
> @rohitgr7 @carmocca thoughts?

I understand, if you want, I'll move out the commits from this PR to another one. But it's more like just moving some of the code pieces somewhere else and then moving them back in follow-up.

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -455,9 +455,22 @@ def __init__(
         self._signal_connector = SignalConnector(self)
         self.tuner = Tuner(self)
 
-        min_steps, max_steps, min_epochs, max_epochs, max_time = _parse_loop_limits(
-            min_steps, max_steps, min_epochs, max_epochs, max_time
+        # Declare attributes to be set in _callback_connector on_trainer_init
+        self._callback_connector.on_trainer_init(
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:459]: I'm not happy that we are changing the order of these. We still have the issue that some of these connectors write back attributes using their internal trainer reference, and the order in which things get initialized here can matter. Have we carefully checked this? 

It looks like all you wanted to do is move `_parse_loop_limit`. Would we be able to move the callback connector and init_debugging_flags and logger connector, back to where they were? For sanity? Or have we done this to resolve another bug?

File: src/pytorch_lightning/trainer/trainer.py [hunk 9]
```diff
@@ -567,66 +563,64 @@ def _init_debugging_flags(
...
 
         self.should_stop = False
         self.state = TrainerState()
         self.num_training_batches = float("inf")
         self.train_dataloader = None
 
-        if num_sanity_val_steps == -1:
-            self.num_sanity_val_steps = float("inf")
-        else:
-            self.num_sanity_val_steps = num_sanity_val_steps
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:628]: I suppose this was an unrelated refactor because you want to classify sanity_val_steps as a debugging flag?

File: src/pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -1138,6 +1132,12 @@ def _run(
         # ----------------------------
         # TRAIN
         # ----------------------------
+        if self.state.fn in (TrainerFn.FITTING, TrainerFn.TUNING):
+            min_epochs, max_epochs = _parse_loop_limits(
+                self.min_steps, self.max_steps, self.min_epochs, self.max_epochs, self
+            )
+            self.fit_loop.min_epochs = min_epochs
+            self.fit_loop.max_epochs = max_epochs
```
Comment (awaelchli) [inline src/pytorch_lightning/trainer/trainer.py:1140]: why aren't we doing it earlier? At this point we've already run model hooks that could have accessed the trainer.
Note, a few lines above we  `verify_loop_configurations(self)`, so I think this should at least move up there

**Review by awaelchli — APPROVED** [review]
Thanks @adamjstewart for working on this and @rohitgr7 / @carmocca for finishing it. 
I think it looks good as a whole solution but the way to get here could have been different IMO. We could have merged the perfectly fine fix that was initially implemented and then iterated on it later. These wouldn't have been conflicting changes and I hope @adamjstewart isn't too upset that we've overridden the direction here. Nonetheless, great work everyone! <3

**Comment by adamjstewart** [comment]
> I hope @adamjstewart isn't too upset that we've overridden the direction here.

Not at all, just glad to see things fixed. My fix was a quick hack, but I agree that a refactor makes sense here.

File: src/pytorch_lightning/trainer/trainer.py (hunk 8 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:459]: Tests should be failing if they aren't moved (I was who initially moved it!). It's very tricky to see these dependencies between connector, the easiest way to double-check would be to move it back and run CI

File: src/pytorch_lightning/trainer/trainer.py (hunk 9 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:628]: I believe I made this change. I moved it because in one of the `_init_debugging_flags`'s branches we were already setting `self.num_sanity_val_steps = 0`, so it made sense that we managed the attribute entirely inside that method

File: src/pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/trainer.py]: done!

File: src/pytorch_lightning/trainer/trainer.py (hunk 8 above)
Comment (rohitgr7) [reply src/pytorch_lightning/trainer/trainer.py:459]: tried moving back and it failed

Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py]: I reverted the re-ordering in [64e4506](https://github.com/Lightning-AI/lightning/pull/13262/commits/64e450663f4e891f2b7f6b3971ad105eef51efe4) and all tests pass. I am still convinced not touching the ordering here is the best way for sanity.

If you still think we should be doing the re-ordering, please feel free revert my last commit, then merge the PR.
